File: src/app/services/websocket.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { BehaviorSubject, Observable, interval, Subscription } from 'rxjs';
import { io, Socket } from 'socket.io-client';
import { environment } from '../../environments/environment';

export enum WebSocketStatus {
  CONNECTING = 'connecting',
  CONNECTED = 'connected',
  DISCONNECTED = 'disconnected',
  ERROR = 'error'
}

@Injectable({
  providedIn: 'root'
})
export class WebSocketService implements OnDestroy {
  private socket: Socket | null = null;
  private status$ = new BehaviorSubject<WebSocketStatus>(WebSocketStatus.DISCONNECTED);
  private messages$ = new BehaviorSubject<any>(null);
  private _connectedAt: Date | null = null;

  // Connection history (last 5 minutes / 300 seconds)
  private _connectionHistory: number[] = new Array(300).fill(0);
  private historyInterval$: Subscription | null = null;

  get connectionStatus$(): Observable<WebSocketStatus> {
    return this.status$.asObservable();
  }

  get currentStatus(): WebSocketStatus {
    return this.status$.value;
  }

  get connectedAt(): Date | null {
    return this._connectedAt;
  }

  get connectionHistory(): number[] {
    return this._connectionHistory;
  }

  get incomingMessages$(): Observable<any> {
    return this.messages$.asObservable();
  }

  connect(): void {
    if (this.socket?.connected) {
      return;
    }

    // Start tracking connection history
    this.startHistoryTracking();

    this.status$.next(WebSocketStatus.CONNECTING);

    const url = this.buildSocketUrl();
    console.log('Socket.io connecting to:', url);

    this.socket = io(url, {
      auth: {
        api_key: environment.apiKey,
        site_id: environment.siteNumber,
        tenant_id: environment.tenantId,
        connection_source: 'app'  // Identify this as a direct app connection (fallback when service not running)
      },
      reconnection: true,
      reconnectionAttempts: 5,
      reconnectionDelay: 3000,
      transports: ['websocket', 'polling']
    });

    this.socket.on('connect', () => {
      console.log('Socket.io connected');
      this._connectedAt = new Date();
      this.status$.next(WebSocketStatus.CONNECTED);
    });

    this.socket.on('disconnect', (reason) => {
      console.log('Socket.io disconnected:', reason);
      this._connectedAt = null;
      this.status$.next(WebSocketStatus.DISCONNECTED);
    });

    this.socket.on('connect_error', (error) => {
      console.error('Socket.io connection error:', error);
      this.status$.next(WebSocketStatus.ERROR);
    });

    this.socket.on('message', (data) => {
      this.messages$.next(data);
    });

    this.socket.onAny((eventName, ...args) => {
      console.log('Socket.io event:', eventName, args);
    });
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
    this.status$.next(WebSocketStatus.DISCONNECTED);
  }

  send(event: string, data?: any): void {
    if (this.socket?.connected) {
      this.socket.emit(event, data);
    } else {
      console.warn('Socket.io is not connected. Cannot send message.');
    }
  }

  on(event: string, callback: (...args: any[]) => void): void {
    this.socket?.on(event, callback);
  }

  off(event: string, callback?: (...args: any[]) => void): void {
    this.socket?.off(event, callback);
  }

  private buildSocketUrl(): string {
    const apiUrl = environment.apiUrl;
    return `${apiUrl}/desktop`;
  }

  private startHistoryTracking(): void {
    if (this.historyInterval$) {
      return; // Already tracking
    }

    this.historyInterval$ = interval(1000).subscribe(() => {
      const isConnected = this.status$.value === WebSocketStatus.CONNECTED ? 1 : 0;
      this._connectionHistory.shift();
      this._connectionHistory.push(isConnected);
    });
  }

  /**
   * Prepopulate connection history from uptime session data
   * This fills in the last 5 minutes of history based on actual connection data
   */
  prepopulateHistory(sessions: Array<{ connected_at: string; disconnected_at: string | null }>): void {
    const now = new Date();
    const fiveMinutesAgo = new Date(now.getTime() - 300 * 1000);

    // Reset history to all zeros first
    this._connectionHistory = new Array(300).fill(0);

    // For each session, mark the connected seconds in our history
    for (const session of sessions) {
      const connectedAt = new Date(session.connected_at);
      // If disconnected_at is null or empty, the session is still active (use current time)
      const disconnectedAt = session.disconnected_at ? new Date(session.disconnected_at) : now;

      // Skip sessions that ended before our 5-minute window
      if (disconnectedAt < fiveMinutesAgo) {
        continue;
      }

      // Calculate the start and end indices in our history array
      // Index 0 = 5 minutes ago, Index 299 = now
      const sessionStart = Math.max(connectedAt.getTime(), fiveMinutesAgo.getTime());
      const sessionEnd = Math.min(disconnectedAt.getTime(), now.getTime());

      const startIndex = Math.floor((sessionStart - fiveMinutesAgo.getTime()) / 1000);
      const endIndex = Math.floor((sessionEnd - fiveMinutesAgo.getTime()) / 1000);

      // Mark all seconds in this range as connected
      for (let i = Math.max(0, startIndex); i <= Math.min(299, endIndex); i++) {
        this._connectionHistory[i] = 1;
      }
    }

    console.log('Connection history prepopulated from', sessions.length, 'sessions');
  }

  ngOnDestroy(): void {
    this.historyInterval$?.unsubscribe();
    this.disconnect();
  }
}

File: service-dotnet/PeiSiteApp/Services/WebSocketService.cs
using PeiSiteApp.Models;
using SocketIOClient;
using SocketIOClient.Transport;

namespace PeiSiteApp.Services;

public class WebSocketService : IDisposable
{
    private SocketIOClient.SocketIO? _socket;
    private readonly object _lock = new();
    private Timer? _historyTimer;
    private volatile bool _stopping;

    private ConnectionStatus _status = ConnectionStatus.Disconnected;
    private DateTime? _connectedAt;
    private int[] _connectionHistory = new int[300];

    public event Action<ConnectionStatus>? StatusChanged;

    public ConnectionStatus Status
    {
        get { lock (_lock) { return _status; } }
    }

    public DateTime? ConnectedAt
    {
        get { lock (_lock) { return _connectedAt; } }
    }

    public int[] ConnectionHistory
    {
        get { lock (_lock) { return (int[])_connectionHistory.Clone(); } }
    }

    public long CurrentUptime
    {
        get
        {
            lock (_lock)
            {
                if (_connectedAt == null || _status != ConnectionStatus.Connected) return 0;
                return (long)(DateTime.UtcNow - _connectedAt.Value).TotalMilliseconds;
            }
        }
    }

    public void Connect(string apiUrl, string apiKey, int siteId, int tenantId)
    {
        if (_socket?.Connected == true) return;

        _stopping = false;
        StartHistoryTracking();
        SetStatus(ConnectionStatus.Connecting);

        var url = $"{apiUrl}/desktop";

        _socket = new SocketIOClient.SocketIO(new Uri(url), new SocketIOClient.SocketIOOptions
        {
            Auth = new
            {
                api_key = apiKey,
                site_id = siteId,
                tenant_id = tenantId,
                connection_source = "app"
            },
            Reconnection = true,
            Transport = TransportProtocol.WebSocket
        });

        _socket.OnConnected += (s, e) =>
        {
            lock (_lock) { _connectedAt = DateTime.UtcNow; }
            SetStatus(ConnectionStatus.Connected);
        };

        _socket.OnDisconnected += (s, reason) =>
        {
            lock (_lock) { _connectedAt = null; }
            SetStatus(ConnectionStatus.Disconnected);
        };

        _socket.OnError += (s, error) =>
        {
            SetStatus(ConnectionStatus.Error);
        };

        _ = Task.Run(async () =>
        {
            while (!_stopping)
            {
                try
                {
                    await _socket.ConnectAsync();
                    break;
                }
                catch
                {
                    SetStatus(ConnectionStatus.Error);
                    await Task.Delay(10000);
                    if (!_stopping) SetStatus(ConnectionStatus.Connecting);
                }
            }
        });
    }

    public void Disconnect()
    {
        _stopping = true;
        StopHistoryTracking();
        if (_socket != null)
        {
            _socket.DisconnectAsync().GetAwaiter().GetResult();
            _socket.Dispose();
            _socket = null;
        }
        SetStatus(ConnectionStatus.Disconnected);
    }

    public void PrepopulateHistory(List<UptimeSession> sessions)
    {
        var now = DateTime.UtcNow;
        var fiveMinutesAgo = now.AddSeconds(-300);

        lock (_lock)
        {
            _connectionHistory = new int[300];

            foreach (var session in sessions)
            {
                if (!DateTime.TryParse(session.ConnectedAt, out var connectedAt)) continue;
                connectedAt = connectedAt.ToUniversalTime();

                var disconnectedAt = session.DisconnectedAt != null
                    && DateTime.TryParse(session.DisconnectedAt, out var da)
                    ? da.ToUniversalTime() : now;

                if (disconnectedAt < fiveMinutesAgo) continue;

                var sessionStart = connectedAt > fiveMinutesAgo ? connectedAt : fiveMinutesAgo;
                var sessionEnd = disconnectedAt < now ? disconnectedAt : now;

                var startIndex = (int)(sessionStart - fiveMinutesAgo).TotalSeconds;
                var endIndex = (int)(sessionEnd - fiveMinutesAgo).TotalSeconds;

                for (int i = Math.Max(0, startIndex); i <= Math.Min(299, endIndex); i++)
                    _connectionHistory[i] = 1;
            }
        }
    }

    private void SetStatus(ConnectionStatus status)
    {
        bool changed;
        lock (_lock)
        {
            changed = _status != status;
            _status = status;
        }
        if (changed) StatusChanged?.Invoke(status);
    }

    private void StartHistoryTracking()
    {
        if (_historyTimer != null) return;
        _historyTimer = new Timer(_ =>
        {
            lock (_lock)
            {
                var isConnected = _status == ConnectionStatus.Connected ? 1 : 0;
                Array.Copy(_connectionHistory, 1, _connectionHistory, 0, 299);
                _connectionHistory[299] = isConnected;
            }
        }, null, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
    }

    private void StopHistoryTracking()
    {
        _historyTimer?.Dispose();
        _historyTimer = null;
    }

    public void Dispose()
    {
        StopHistoryTracking();
        _socket?.Dispose();
    }
}
